File: benchmarks/runners/node/coldstart.mjs
import { spawn } from "node:child_process";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { casesForMetric } from "../../lib/manifest.mjs";
import { computeStats } from "../../lib/stats.mjs";

const __dirname = dirname(fileURLToPath(import.meta.url));
const CHILD = join(__dirname, "coldstart-child.mjs");

/**
 * Spawn one fresh process and measure true cold-start: wall-clock from just
 * before spawn to the child's "READY" marker (process launch + module/addon load
 * + isolate init). first-run is timed in-process by the child. Rejects on a
 * non-zero exit or a missing READY/JSON line so a failed sample never records a
 * bogus timing.
 */
function sampleOnce(corpusDir, caseId) {
  return new Promise((resolve, reject) => {
    const t0 = process.hrtime.bigint();
    const child = spawn(process.execPath, [CHILD, corpusDir, caseId], {
      stdio: ["ignore", "pipe", "pipe"],
    });
    let coldStartMs;
    let stdout = "";
    let stderr = "";
    child.stdout.setEncoding("utf-8");
    child.stdout.on("data", (chunk) => {
      stdout += chunk;
      if (coldStartMs === undefined && stdout.includes("READY\n")) {
        // Stamp the moment the runtime reported ready. Everything before this
        // line (launch, load, init) is the cold-start cost.
        coldStartMs = Number(process.hrtime.bigint() - t0) / 1e6;
      }
    });
    child.stderr.setEncoding("utf-8");
    child.stderr.on("data", (chunk) => (stderr += chunk));
    child.on("error", reject);
    child.on("close", (code) => {
      if (code !== 0) {
        reject(new Error(`coldstart child failed for '${caseId}' (exit ${code})\n${stderr}`));
        return;
      }
      if (coldStartMs === undefined) {
        reject(new Error(`coldstart child for '${caseId}' never printed READY\n${stderr}`));
        return;
      }
      const jsonLine = stdout.split("\n").filter((l) => l && l !== "READY").pop();
      if (!jsonLine) {
        reject(new Error(`coldstart child for '${caseId}' printed no result line\n${stderr}`));
        return;
      }
      const { firstRunMs } = JSON.parse(jsonLine);
      resolve({ coldStartMs, firstRunMs });
    });
  });
}

/**
 * @returns {Promise<Array<{id,metric,unit,stats,iterations}>>}
 */
export async function runColdStartAndFirstRun(manifest, { samplesOverride } = {}) {
  const rows = [];
  const ids = new Set([
    ...casesForMetric(manifest, "cold-start").map((c) => c.id),
    ...casesForMetric(manifest, "first-run").map((c) => c.id),
  ]);

  for (const id of ids) {
    const c = manifest.cases.find((x) => x.id === id);
    const n = samplesOverride ?? c.iterations?.samples ?? 20;
    const colds = [];
    const firsts = [];
    for (let i = 0; i < n; i++) {
      const { coldStartMs, firstRunMs } = await sampleOnce(manifest.corpusDir, id);
      colds.push(coldStartMs);
      firsts.push(firstRunMs);
    }
    if (c.metrics.includes("cold-start")) {
      rows.push({ id, metric: "cold-start", unit: "ms", stats: computeStats(colds), iterations: n });
    }
    if (c.metrics.includes("first-run")) {
      rows.push({ id, metric: "first-run", unit: "ms", stats: computeStats(firsts), iterations: n });
    }
  }
  return rows;
}

File: benchmarks/runners/python/coldstart.py
"""Spawn orchestrator: N fresh processes per case, aggregate cold-start /
first-run rows. Mirrors runners/node/coldstart.mjs."""

import json
import subprocess
import sys
import time
from pathlib import Path

from manifest import cases_for_metric
from stats import compute_stats

_CHILD = str(Path(__file__).resolve().parent / "coldstart_child.py")


def _spawn_sample(corpus_dir, case_id):
    """Spawn one fresh process and measure true cold-start: wall-clock from just
    before spawn to the child's "READY" marker (process launch + interpreter
    start + dlopen(dwlib) + isolate creation). first-run is timed in-process by
    the child. Raises on a non-zero exit or a missing READY/JSON line so a failed
    sample never records a bogus timing. Returns (coldStartMs, firstRunMs)."""
    t0 = time.perf_counter_ns()
    p = subprocess.Popen(
        [sys.executable, _CHILD, str(corpus_dir), case_id],
        stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True,
    )
    cold_start_ms = None
    result_line = None
    for line in p.stdout:
        line = line.rstrip("\n")
        if line == "READY":
            # Stamp the moment the runtime reported ready; everything before this
            # (launch, dlopen, init) is the cold-start cost.
            cold_start_ms = (time.perf_counter_ns() - t0) / 1e6
        elif line:
            result_line = line
    stderr = p.stderr.read()
    code = p.wait()
    if code != 0:
        raise RuntimeError(f"coldstart child failed for '{case_id}' (exit {code})\n{stderr}")
    if cold_start_ms is None:
        raise RuntimeError(f"coldstart child for '{case_id}' never printed READY\n{stderr}")
    if not result_line:
        raise RuntimeError(f"coldstart child for '{case_id}' printed no result line\n{stderr}")
    return cold_start_ms, json.loads(result_line)["firstRunMs"]


def run_cold_start_and_first_run(manifest, sample_fn=_spawn_sample, samples_override=None):
    # Unique case ids that declare cold-start and/or first-run, preserving order.
    ids = []
    for c in cases_for_metric(manifest, "cold-start") + cases_for_metric(manifest, "first-run"):
        if c["id"] not in ids:
            ids.append(c["id"])

    rows = []
    for cid in ids:
        c = next(x for x in manifest["cases"] if x["id"] == cid)
        n = samples_override if samples_override is not None \
            else c.get("iterations", {}).get("samples", 20)
        colds, firsts = [], []
        for _ in range(n):
            cold_ms, first_ms = sample_fn(manifest["corpusDir"], cid)
            colds.append(cold_ms)
            firsts.append(first_ms)
        if "cold-start" in c["metrics"]:
            rows.append({"id": cid, "metric": "cold-start", "unit": "ms",
                         "stats": compute_stats(colds), "iterations": n})
        if "first-run" in c["metrics"]:
            rows.append({"id": cid, "metric": "first-run", "unit": "ms",
                         "stats": compute_stats(firsts), "iterations": n})
    return rows
